### PR TITLE
[WIP] Sets: Fix relational between Cartesian Product and Finite Sets

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

1. `is_subset` gives a wrong result
```python
>>> from sympy import *

>>> a = FiniteSet(1, 2)
>>> b = ProductSet(a, a)
>>> c = FiniteSet((1, 1), (1, 2), (2, 1), (2, 2))
>>> b.intersection(c) == c.intersection(b)
True
>>> b.is_subset(c)
False
>>> c.is_subset(b)
True
```

2. Calling `Eq.simplify` for sets gives `AttributeError`
```python
>>> Eq(b, c).simplify()
```
```AttributeError: 'ProductSet' object has no attribute 'simplify'```

3. Rewriting `ProductSet` to `FiniteSet` does not work
```python
>>> b.rewrite(FiniteSet)
{1, 2} x {1, 2}
```

I think there have to be some assumption handler for detecting if the set is a finite set, instead of using `is_iterable`. I see `is_iterable` is being used in some infinite sets like `Integers`.
I may consider `is_FiniteSet`, but is it intended for use in other place than having strict assumption.

Also, there had to be things modified for core Relational class

#### Other comments

- Made line spacings consistent in the module

#### Release Notes

- core
  - Fixed `Eq.simplify()` not working with relationals involving sets. 
- sets
  - Added `simplify` method for sets.
  - Fixed `is_subset` giving wrong result between `FiniteSet` and its `ProductSet`
  - Added rewriting method for `ProductSet` to `FiniteSet`.

### CODE REVIEW

File: sympy/core/relational.py [hunk 1]
```diff
@@ -281,20 +281,40 @@ def equals(self, other, failing_expression=False):
 
     def _eval_simplify(self, ratio, measure, rational, inverse):
         r = self
-        r = r.func(*[i.simplify(ratio=ratio, measure=measure,
-                                rational=rational, inverse=inverse)
-                     for i in r.args])
+
+        # XXX Some sympy classes like Set does not have 'simplify' attribute.
+        def try_simp(i):
```
Comment (oscarbenjamin) [inline sympy/core/relational.py]: Can we add the simplify method to sets if that is what is needed? It seems wrong to add this here

File: sympy/sets/sets.py [hunk 2]
```diff
@@ -686,6 +716,36 @@ def is_iterable(self):
         """
         return all(set.is_iterable for set in self.sets)
 
+    @property
+    def is_FiniteSet(self):
```
Comment (oscarbenjamin) [inline sympy/sets/sets.py]: I would expect a capitalised name like this to mean `isinstance(self, FiniteSet)`. Maybe `is_finite_set` is better.

File: sympy/sets/sets.py [hunk 3]
```diff
@@ -686,6 +716,36 @@ def is_iterable(self):
...
+
+        Only the lemma that the finite Cartesian product of finite sets is a
+        finite set is implemented.
+
+        References
+        ==========
+
+        .. [1] https://proofwiki.org/wiki/Product_of_Finite_Sets_is_Finite
+        """
+        if len(self.args) < S.Infinity:
```
Comment (oscarbenjamin) [inline sympy/sets/sets.py]: If args is a tuple then it will always have finite length

File: sympy/sets/sets.py [hunk 4]
```diff
@@ -735,6 +768,35 @@ def is_iterable(self):
...
+        finite set is implemented.
+
+        References
+        ==========
+
+        .. [1] https://proofwiki.org/wiki/Product_of_Finite_Sets_is_Finite
+        """
+        if all(arg.is_finite_set for arg in self.args):
+            return True
+        return False
```
Comment (oscarbenjamin) [inline sympy/sets/sets.py]: I guess that some sets could have `is_finite_set` as None e.g.:
```julia
In [6]: a, x = symbols('a, x')

In [7]: C = ConditionSet(x, Eq(a*x,0), S.Reals)                                                                                                                                                                    

In [8]: C
Out[8]: {x | x ∊ ℝ ∧ a⋅x = 0}

In [9]: C.subs(a, 1) # should simplify to {0}
Out[9]: {x | x ∊ ℝ ∧ x = 0}

In [10]: C.subs(a, 0)
Out[10]: ℝ
```
If this was ternary logic then it should be something like:
```
fuzzy_and(arg.is_finite_set for arg in self.args)
```
Might also need to check if any arg is the empty set.

**Comment by sylee957** [comment]
If it is going to use tenary logic, default should be None instead of False though.

**Comment by oscarbenjamin** [comment]
>  default should be None

Do you mean in class Set? If so then yes I agree.

We should add it as False in Integers, Reals etc. though

**Comment by oscarbenjamin** [comment]
Interesting that the doctest failure here is
```
Failed example:
    pprint(nonlinsolve(system, [a, b, c, d]), use_unicode=False)
Expected:
      -1       1               1      -1
    {(---, -d, -, {d} \ {0}), (-, -d, ---, {d} \ {0})}
       d       d               d       d
Got:
      -1       1       1      -1      
    {(---, -d, -, d), (-, -d, ---, d)}
       d       d       d       d      
```
I had trouble with this test here: https://github.com/sympy/sympy/pull/16666#discussion_r279164891

**Comment by oscarbenjamin** [comment]
@sylee957 what's the status of this?

File: sympy/sets/fancysets.py [hunk 5]
```diff
@@ -339,6 +341,10 @@ def __new__(cls, flambda, *sets):
     lamda = property(lambda self: self.args[0])
     base_sets = property(lambda self: self.args[1:])
 
+    @property
+    def is_empty(self):
+        return fuzzy_and([s.is_empty for s in self.base_sets])
```
Comment (oscarbenjamin) [inline sympy/sets/fancysets.py]: This should be `fuzzy_or`

**Comment by czgdp1807** [comment]
@sylee957 Are you still working on it? It looks like one of the requirements of the PR has been satisfied by #17848 but I am not sure about others. Please update. Thanks.

**Comment by sylee957** [comment]
I have stalled this PR for a while because it needed some changes in the #17848, but I'll rebase it and see if this needs some minor tweaks or some major reworks.

**Comment by czgdp1807** [comment]
@sylee957 Current status on `master`,
```python
>>> a = FiniteSet(1, 2)
>>> b = ProductSet(a, a)
>>> c = FiniteSet((1, 1), (1, 2), (2, 1), (2, 2))
>>> b.intersection(c) == c.intersection(b)
True
>>> b.is_subset(c)
>>> c.is_subset(b)
True
>>> Eq(b, c).simplify()
Traceback (most recent call last):
  File "/usr/lib/python3.6/code.py", line 91, in runcode
    exec(code, self.locals)
  File "<console>", line 1, in <module>
  File "/home/czgdp1807ssd/sympy_project/sympy/sympy/core/basic.py", line 1655, in simplify
    return simplify(self, **kwargs)
  File "/home/czgdp1807ssd/sympy_project/sympy/sympy/simplify/simplify.py", line 559, in simplify
    return _eval_simplify(**kwargs)
  File "/home/czgdp1807ssd/sympy_project/sympy/sympy/core/relational.py", line 646, in _eval_simplify
    e = super(Equality, self)._eval_simplify(**kwargs)
  File "/home/czgdp1807ssd/sympy_project/sympy/sympy/core/relational.py", line 308, in _eval_simplify
    elif dif.equals(0):  # XXX this is expensive
AttributeError: 'Complement' object has no attribute 'equals'
>>> b.rewrite(FiniteSet)
      2
{1, 2} 
>>> 
```

**Comment by czgdp1807** [comment]
Probably this PR can be closed and a new issue can be opened as the issue has changed.

**Pull request closed without merging**